src/ITensorNetworks/MPScalculus/contract.jl

```diff
@@ -0,0 +1,54 @@
+using ChainRulesCore
+
+struct MPSTensor
+  mps::MPS
+  cutoff::Float64
+  maxdim::Integer
+end
+
+AbstractTensor = Union{ITensor,MPSTensor}
+
+function ITensors.contract(mps1::MPS, mps2::MPS; cutoff=1e-15, maxdim=1000)
+  ## TODO: modify this function based on https://arxiv.org/pdf/1912.03014.pdf
+  tensor = contract(vcat(collect(mps1), collect(mps2))...)
+  return if size(tensor) == ()
+    MPS([tensor])
+  else
+    MPS(tensor, inds(tensor); cutoff=cutoff, maxdim=maxdim)
+  end
+end
+
+MPSTensor(mps::MPS) = MPSTensor(mps, 1e-15, 1000)
+
+function ITensors.inds(tensor::MPSTensor)
+  return siteinds(tensor.mps) == [nothing] ? [] : siteinds(tensor.mps)
+end
+
+# contract into one ITensor
+ITensors.ITensor(t::MPSTensor) = contract(collect(t.mps)...)
+
+ITensors.contract(t1::MPSTensor) = t1
+
+function ITensors.contract(t1::MPSTensor, t2::MPSTensor)
+  mps_out = contract(t1.mps, t2.mps; cutoff=t1.cutoff, maxdim=t1.maxdim)
```

Comment:
I suppose this points to an ambiguity with storing the `cutoff` and `maxdim` inside the `MPSTensor`, since then there is a question about which ones take precedence (we could be conservative and use `min(t1.cutoff, t2.cutoff)` and `max(t1.maxdim, t2.maxdim)`, but I think the issue of ambiguity is still there).

Is there such a disadvantage to just pass `maxdim` and `cutoff` as keyword arguments? It is pretty easy to forward them from high level functions to low level functions, it just seems like it is a matter of adding them as keyword arguments to the higher level network contraction functions. This is what we do throughout the ITensor and it seems to work fine. Then there is only one `maxdim` and `cutoff` for the entire approximate contraction, which I think is what you want anyway.

Reply:
The idea of storing these parameters insode `MPSTensor` is to make the AD system general. The contract function the AD system called, https://github.com/ITensor/ITensorNetworkAD.jl/blob/main/src/ITensorAutoHOOT/utils.jl#L46, should not contain any keyword arguments specific to any approximation algorithm, since that will make the system not general. For example, if we introduce another new iterative based approximate contraction scheme, then there might be another parameter to show the number of iterations. This information should not be exposed to the AD system in my eyes.

 And this is the reason I put all these parameters inside MPSTensor, since these parameters will be carried around all the time.

Reply:
From what I can tell, all you would really need to do is add `; kwargs...` to the function signatures, such as change:
```julia
function compute_graph!(out_nodes, node_dict)
```
to:
```julia
function compute_graph!(out_nodes, node_dict; kwargs...)
```
and then:
```julia
      node_dict[node] = contract(input_list)
```
to:
```julia
      node_dict[node] = contract(input_list; kwargs...)
```
This would allow any arbitrary options to get forwarded in the future (such as an iteration parameter).

This is what we do in our general approximate gate evolution function in ITensors.jl (the top level function is [here](https://github.com/ITensor/ITensors.jl/blob/main/src/mps/abstractmps.jl#L1852-L1866)). The `kwargs` get forwarded down through many layers of code where they eventually get used when performing the approximate contraction of the gate with the MPS/MPO. We could swap out the approximate algorithm we use and forward different arguments, but the top level function never has to change.

Reply:
I see, that makes sense. Then I guess we don't need the MPSTensor struct, can direct define `contract`, `ITensor`, and `getindex` on MPS.

Reply:
Possibly, though I think we want to think carefully about the definitions to ensure they make sense in general for the `MPS` type.

1. `contract`: I believe the definition for `contract(::MPS, ::MPS)` should be fine. I was thinking of defining that operation for when the `MPS` have all site indices in common, in which case it reduces to a scalar value, but we can have that as a special case of the more general approximate contraction that you have in mind.
2. `ITensor`: I think your definition of contracting all of the tensors of the MPS together makes sense as a definition.
3. `inds`: If I were to define `inds(::MPS)` in general, I would define it as something like `inds(M::MPS) = noncommoninds(M...)`, i.e. the concatenation of the site indices into a single collection of indices. But I think that could be up for interpretation. What is the use case for `inds(::MPS)` here?
4. `getindex`: I think this one is the most ambiguous. Currently, indexing into an MPS extracts a tensor of the MPS, not an element of the MPS when the MPS is treated as a large tensor. So based on the current definition of `getindex`, I would think that `M[]` would mean getting the first ITensor of the MPS, if the MPS has length 1, just like it means for a collection of Julia Arrays:
```julia
julia> v = [randn(2, 2)]
1-element Vector{Matrix{Float64}}:
 [-1.5101394161681398 -0.015271989173999531; -0.09443655629156805 -1.1668103862767556]

julia> v[]
2×2 Matrix{Float64}:
 -1.51014    -0.015272
 -0.0944366  -1.16681
```
This is a case where it may be best to use a different function, like `scalar(::MPS)`, which would be less ambiguous.

Reply:
I see. `inds(Tensor)`, where Tensor can be ITensor or MPSTensor, is used to get the dimension information of the tensor and then map the tensor to the tensor in AutoHOOT. I guess the `noncommoninds(M...)` is the same as `siteinds(M)`, is that right?

I can see where the `getindex` confusion comes from. Since here we really want the MPS to represent a single tensor, we want the `getindex` to extract the elements, so that the definition is consistent with ITensor.  

Maybe it's OK to keep the MPSTensor struct for now, let me know what you think

Reply:
> I see. `inds(Tensor)`, where Tensor can be ITensor or MPSTensor, is used to get the dimension information of the tensor and then map the tensor to the tensor in AutoHOOT. I guess the `noncommoninds(M...)` is the same as `siteinds(M)`, is that right?

In general they are nearly the same, but in some cases an MPS can have multiple site indices per tensor, so for example `siteinds` may output something like `[[i1, j1], [i2, j2], [i3, j3], ...]` while `noncommoninds(M...)` would return the flattened set of indices `[i1, j1, i2, j2, i3, j3, ...]`, so the question would be about what format `inds` should output. I would lean towards `inds(::MPS)` returning the flattened data structure (this sounds more consistent with what is needed by `AutoHOOT` anyway).

> 
> I can see where the `getindex` confusion comes from. Since here we really want the MPS to represent a single tensor, we want the `getindex` to extract the elements, so that the definition is consistent with ITensor.

Right. It is kind of a tension between whether an `MPS` acts as a collection of tensors or as tensor-like object that happens to be in a certain tensor network form. I think with the right generic functions (like using a generic `scalar` function), we can make the `MPS` object act more like an `ITensor`, but it requires some thinking about the design.

> 
> Maybe it's OK to keep the MPSTensor struct for now, let me know what you think

Yeah, we could just keep `MPSTensor` for now, and think about how to design things to be more generic between the `MPS` and `ITensor` types so it is not needed in the future.

This bumps against questions I have had in general about how much a tensor network should act like an `ITensor`, for example `inds(tn::ITensorNetwork) = noncommoninds(tn...)` would be a good generic definition for a future `ITensorNetwork` type. I think many index and contraction related functions for `ITensor`s could be generalized to an `ITensorNetwork`.